# Flight Operations Data Analysis

#### Objective

The objective of this analysis is to explore the Flight Operations dataset using Pandas and identify the patterns and key findings through data inspection, cleaning, selection, filtering, sorting, grouping, aggregation, and transformation.

##### Importing Required Libraries

Pandas is imported to perform data manipulation, analysis, and transformation on the Flight Operations dataset.

In [1]:
import pandas as pd
print("Pandas imported successfully!")

Pandas imported successfully!


#### Loading the Dataset

The Flight Operations dataset is loaded into a Pandas DataFrame for further analysis. A DataFrame allows us to inspect, manipulate, filter, and analyze the flight operation records efficiently.

In [2]:
df = pd.read_csv("Flight_Operations_Dataset.csv")

#### Understanding the Dataset

Before performing the analysis, it is important to understand the structure and contents of the dataset. In this section, we inspect the first few records, dataset dimensions, column names, data types, missing values, and descriptive statistics.

##### Displaying the First Five Records

The `head()` function is used to display the first five rows of the dataset and get an initial understanding of the available data.

In [3]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


##### Checking the Dataset Shape

The `shape` attribute is used to determine the number of rows and columns in the dataset. This provides a quick overview of the dataset size.

In [4]:
df.shape

(180, 17)

##### Viewing Column Names

The `columns` attribute is used to display the names of all columns in the dataset. This helps us understand the variables available for further analysis.

In [5]:
df.columns

Index(['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination',
       'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity',
       'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather',
       'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference',
       'Passenger_Satisfaction'],
      dtype='str')

##### Inspecting Dataset Information

The `info()` method provides information about the dataset, including the column names, data types, number of non-null values, and memory usage. It also helps identify missing values.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    str    
 1   Flight_Date             180 non-null    str    
 2   Airline                 180 non-null    str    
 3   Origin                  180 non-null    str    
 4   Destination             180 non-null    str    
 5   Aircraft                180 non-null    str    
 6   Travel_Class            180 non-null    str    
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    str    
 12  Weather                 180 non-null    str    
 13  Booking_Channel         180 non-null    str    
 14  Avg_Baggage_Kg          180 non-null    int64  
 15  

##### Checking Missing Values

Missing values can affect the accuracy of the analysis. The `isnull().sum()` operation is used to identify the number of missing values in each column.

In [7]:
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

##### Generating Descriptive Statistics

The `describe()` method is used to generate summary statistics for the numerical columns in the dataset, such as count, mean, standard deviation, minimum, maximum, and quartile values.

In [8]:
df.describe()

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


#### Data Cleaning and Missing Value Handling

Before performing further analysis, missing values need to be handled to ensure that the dataset is suitable for analysis. The missing values in the `Delay_Minutes` column are replaced using the median value of that column.

In [10]:
# Check missing values before handling
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             7
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

In [11]:
# Fill missing values
df["Delay_Minutes"] = df["Delay_Minutes"].fillna(df["Delay_Minutes"].median())

After handling the missing values, the dataset is checked again to confirm that no missing values remain.

In [12]:
df.isnull().sum()

Flight_ID                 0
Flight_Date               0
Airline                   0
Origin                    0
Destination               0
Aircraft                  0
Travel_Class              0
Passengers                0
Seat_Capacity             0
Average_Ticket_Price      0
Delay_Minutes             0
Flight_Status             0
Weather                   0
Booking_Channel           0
Avg_Baggage_Kg            0
Meal_Preference           0
Passenger_Satisfaction    0
dtype: int64

#### Data Selection and Filtering

Data selection and filtering are used to focus on specific columns and records that are relevant to the analysis.

##### Selecting Relevant Columns

Relevant columns are selected to examine important information such as the airline, destination, number of passengers, flight delay, and passenger satisfaction.

In [15]:
df[[
    "Airline",
    "Origin",
    "Destination",
    "Passengers",
    "Delay_Minutes",
    "Passenger_Satisfaction"
]].head()

,Airline,Origin,Destination,Passengers,Delay_Minutes,Passenger_Satisfaction
0,SpiceJet,Mumbai,Bengaluru,210,50.0,3
1,Akasa Air,Delhi,Mumbai,70,0.0,5
2,Air India,Kochi,Delhi,67,5.0,2
3,Akasa Air,Srinagar,Delhi,153,35.0,2
4,Air India,Pune,Delhi,178,20.0,5


##### Selecting a Single Column

A single column can be selected to examine the values of a particular variable. The following operation displays the number of passengers for the first few flights.

In [16]:
df["Passengers"].head()

0    210
1     70
2     67
3    153
4    178
Name: Passengers, dtype: int64

##### Filtering Delayed Flights

Flights with a delay greater than zero minutes are filtered to examine the records of delayed flights.

In [17]:
delayed_flights = df[df["Delay_Minutes"] > 0]

delayed_flights.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5
5,FL0006,2026-05-20,Akasa Air,Kochi,Delhi,Airbus A319,Economy,123,220,5636,35.0,Delayed,Cloudy,Travel Agency,13,Vegetarian,2


##### Filtering Flights Based on Passenger Satisfaction

The passenger satisfaction column is examined to identify the available satisfaction categories before applying a filter.

In [18]:
df["Passenger_Satisfaction"].unique()

array([3, 5, 2, 4])

##### Filtering Flights with More Than 200 Passengers

Flights carrying more than 200 passengers are filtered to focus on high-passenger flight operations.

In [19]:
high_passenger_flights = df[df["Passengers"] > 200]

high_passenger_flights.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
69,FL0070,2026-06-12,SpiceJet,Srinagar,Delhi,Boeing 737,Economy,201,189,2959,20.0,Delayed,Rain,Online Travel Portal,12,Vegan,4
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,0.0,On Time,Clear,Travel Agency,8,Vegan,2
148,FL0149,2026-02-10,Akasa Air,Jaipur,Mumbai,Airbus A319,Economy,202,210,6239,5.0,On Time,Rain,Online Travel Portal,8,No Meal,5


##### Filtering Using Multiple Conditions

Multiple conditions can be combined to select specific flight records. The following operation filters flights with more than 200 passengers and a delay of more than 30 minutes.

In [20]:
filtered_flights = df[
    (df["Passengers"] > 200) &
    (df["Delay_Minutes"] > 30)
]

filtered_flights.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4


#### Sorting the Data

Sorting helps organize the dataset based on specific columns. In this section, the flight records are sorted based on delay duration and passenger count to identify notable patterns.

##### Sorting Flights by Delay Duration

The dataset is sorted in descending order of delay minutes to identify the flights with the highest delays.

In [21]:
df.sort_values(by="Delay_Minutes", ascending=False).head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3


##### Sorting Flights by Number of Passengers

The dataset is sorted in descending order based on the number of passengers to identify the flights carrying the highest number of passengers.

In [22]:
df.sort_values(by="Passengers", ascending=False).head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
92,FL0093,2026-04-30,Air India Express,Srinagar,Delhi,Airbus A319,Business,209,220,12970,0.0,On Time,Clear,Travel Agency,8,Vegan,2
148,FL0149,2026-02-10,Akasa Air,Jaipur,Mumbai,Airbus A319,Economy,202,210,6239,5.0,On Time,Rain,Online Travel Portal,8,No Meal,5
69,FL0070,2026-06-12,SpiceJet,Srinagar,Delhi,Boeing 737,Economy,201,189,2959,20.0,Delayed,Rain,Online Travel Portal,12,Vegan,4


#### Grouping and Aggregation

Grouping and aggregation are used to analyze flight operations based on different categories. These operations help identify patterns by calculating summary values such as average delays, total passengers, and the number of flights.

##### Average Delay by Airline

The average delay for each airline is calculated to compare the operational performance of different airlines.

In [23]:
average_delay_by_airline = df.groupby("Airline")["Delay_Minutes"].mean()

average_delay_by_airline.sort_values(ascending=False)

Airline
IndiGo               25.031250
SpiceJet             22.343750
Air India            20.583333
Vistara              17.891892
Akasa Air            14.096774
Air India Express    11.291667
Name: Delay_Minutes, dtype: float64

##### Total Passengers by Airline

The total number of passengers is calculated for each airline to identify which airline carried the highest number of passengers.

In [24]:
total_passengers_by_airline = df.groupby("Airline")["Passengers"].sum()

total_passengers_by_airline.sort_values(ascending=False)

Airline
Vistara              4484
IndiGo               4158
SpiceJet             3929
Akasa Air            3427
Air India            3074
Air India Express    3027
Name: Passengers, dtype: int64

##### Number of Flights by Destination

The number of flight records for each destination is calculated to identify the most frequently occurring destinations in the dataset.

In [25]:
flights_by_destination = df.groupby("Destination").size()

flights_by_destination.sort_values(ascending=False)

Destination
Delhi        51
Mumbai       30
Bengaluru    28
Kolkata      20
Chennai      17
Hyderabad    13
Srinagar     13
Kochi         8
dtype: int64

#### Data Transformation

Data transformation is used to create or modify columns so that the dataset can provide additional insights. In this section, new columns are created based on existing flight operation data.

##### Creating a Delay Category

A new column named `Delay_Category` is created to classify flights based on their delay duration.

In [26]:
def categorize_delay(delay):
    if delay == 0:
        return "On Time"
    elif delay <= 30:
        return "Low Delay"
    else:
        return "High Delay"

df["Delay_Category"] = df["Delay_Minutes"].apply(categorize_delay)

df[["Delay_Minutes", "Delay_Category"]].head()

,Delay_Minutes,Delay_Category
0,50.0,High Delay
1,0.0,On Time
2,5.0,Low Delay
3,35.0,High Delay
4,20.0,Low Delay


##### Creating a Passenger Category

A new column named `Passenger_Category` is created to classify flights based on the number of passengers.

In [27]:
def categorize_passengers(passengers):
    if passengers < 150:
        return "Low"
    elif passengers <= 250:
        return "Medium"
    else:
        return "High"

df["Passenger_Category"] = df["Passengers"].apply(categorize_passengers)

df[["Passengers", "Passenger_Category"]].head()

,Passengers,Passenger_Category
0,210,Medium
1,70,Low
2,67,Low
3,153,Medium
4,178,Medium


#### Analysis

The analysis helps identify important patterns and trends in flight operations, such as airline delays, passenger traffic, and destination frequency.

##### Airline with the Highest Average Delay

The average delay is calculated for each airline to identify which airline has the highest average flight delay.

In [28]:
average_delay_by_airline = df.groupby("Airline")["Delay_Minutes"].mean()

average_delay_by_airline.sort_values(ascending=False)

Airline
IndiGo               25.031250
SpiceJet             22.343750
Air India            20.583333
Vistara              17.891892
Akasa Air            14.096774
Air India Express    11.291667
Name: Delay_Minutes, dtype: float64

In [29]:
average_delay_by_airline.sort_values(ascending=False).head(1)

Airline
IndiGo    25.03125
Name: Delay_Minutes, dtype: float64

*IndiGo has the highest average delay of approximately 25.03 minutes among the airlines in the dataset. This indicates that IndiGo experienced comparatively higher average flight delays during the period represented by the dataset.*

##### Airline with the Highest Number of Passengers

The total number of passengers is calculated for each airline to identify which airline handled the highest number of passengers in the dataset.

In [30]:
total_passengers_by_airline = df.groupby("Airline")["Passengers"].sum()

total_passengers_by_airline.sort_values(ascending=False)

Airline
Vistara              4484
IndiGo               4158
SpiceJet             3929
Akasa Air            3427
Air India            3074
Air India Express    3027
Name: Passengers, dtype: int64

*Vistara has the highest total number of passengers with 4,484 passengers in the dataset. This indicates that Vistara handled the highest passenger volume among the airlines included in the dataset.*

##### Most Frequent Destination

The number of flights is calculated for each destination to identify which destination appears most frequently in the dataset.

In [31]:
flights_by_destination = df.groupby("Destination").size()

flights_by_destination.sort_values(ascending=False)

Destination
Delhi        51
Mumbai       30
Bengaluru    28
Kolkata      20
Chennai      17
Hyderabad    13
Srinagar     13
Kochi         8
dtype: int64

In [32]:
flights_by_destination.sort_values(ascending=False).head(1)

Destination
Delhi    51
dtype: int64

*Delhi is the most frequent destination in the dataset, with 51 flight records. This indicates that Delhi had the highest number of flights among the destinations included in the dataset.*

##### Distribution of Flights by Delay Category

The number of flights in each delay category is calculated to understand the overall distribution of on-time and delayed flights.

In [33]:
delay_category_counts = df["Delay_Category"].value_counts()

delay_category_counts

Delay_Category
On Time       70
Low Delay     70
High Delay    40
Name: count, dtype: int64

*The dataset contains 70 on-time flights, 70 flights with low delays, and 40 flights with high delays. This shows that on-time and low-delay flights are equally common, while high-delay flights are less frequent.*

##### Distribution of Flights by Passenger Category

The number of flights in each passenger category is calculated to understand the distribution of flights based on passenger volume.

In [34]:
passenger_category_counts = df["Passenger_Category"].value_counts()

passenger_category_counts

Passenger_Category
Low       120
Medium     60
Name: count, dtype: int64

*The dataset contains 120 flights in the Low passenger category and 60 flights in the Medium passenger category. No flights fall into the High passenger category based on the defined classification. This indicates that most flights in the dataset carried fewer than 150 passengers.*

##### Average Passengers by Delay Category

The average number of passengers is calculated for each delay category to examine whether passenger volume varies with flight delay levels.

In [35]:
average_passengers_by_delay = df.groupby("Delay_Category")["Passengers"].mean()

average_passengers_by_delay

Delay_Category
High Delay    123.325000
Low Delay     123.900000
On Time       121.328571
Name: Passengers, dtype: float64

*The average passenger count is highest for low-delay flights at approximately 123.90 passengers, followed by high-delay flights at approximately 123.33 passengers. On-time flights have the lowest average passenger count at approximately 121.33 passengers. The differences are relatively small, indicating that passenger volume does not vary significantly across the delay categories in this dataset.*

### Key Observations

Based on the analysis performed on the Flight Operations dataset, the following key observations were identified:

1. **IndiGo had the highest average delay**, with an average delay of approximately 25.03 minutes among the airlines analyzed.

2. **Vistara recorded the highest total passenger count**, with 4,484 passengers in the dataset.

3. **Delhi was the most frequent destination**, appearing in 51 flight records, making it the most commonly occurring destination in the dataset.

4. **On-time and low-delay flights were equally common**, with 70 flights in each category, while 40 flights were classified as high-delay flights.

5. **Most flights belonged to the Low passenger category**, with 120 flights, while 60 flights belonged to the Medium passenger category. No flights were classified as High passenger category.

6. **The average passenger count was highest for Low Delay flights**, at approximately 123.90 passengers, while On Time flights had the lowest average at approximately 121.33 passengers.

7. **Passenger volume showed only a small difference across delay categories**, suggesting that the number of passengers did not vary significantly with the delay category in this dataset.

8. **Airline performance varied in terms of average delay**, with IndiGo having the highest average delay and Air India Express having the lowest average delay at approximately 11.29 minutes.

### Conclusion

The Flight Operations dataset was analyzed using a complete Pandas data analysis workflow. The process included loading and understanding the dataset, inspecting its structure and statistics, selecting and filtering data, sorting records, grouping and aggregating data, and creating new categorical columns through data transformation.

The analysis revealed useful patterns in airline delays, passenger volumes, destinations, and delay categories. IndiGo had the highest average delay, while Vistara handled the highest total number of passengers. Delhi was the most frequently recorded destination. The analysis also showed that passenger volume had only a small variation across different delay categories.

Overall, the analysis demonstrates how Pandas can be used to organize, transform, summarize, and extract analytical findings from flight operations data.